File: Parallel-Csharp/src/Parallel/Linq101/Restrictions.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Parallel.Linq101.DataSources;

namespace Parallel.Linq101;

public class Restrictions
{
    public List<Product> GetProductList() => Products.ProductList;

    public List<Customer> GetCustomerList() => Customers.CustomerList;

    public int LowNumbers()
    {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };

        var lowNums = numbers.Where(n => n < 5).ToArray();
        Console.WriteLine("Numbers < 5:");
        foreach (var x in lowNums)
            Console.WriteLine(x);
        return 0;
    }

    public int ProductsOutOfStock()
    {
        List<Product> products = GetProductList();

        var soldOutProducts = products.Where(p => p.UnitsInStock == 0).ToList();

        Console.WriteLine("Sold out products:");
        foreach (var product in soldOutProducts)
            Console.WriteLine($"{product.ProductName} is sold out!");

        return 0;
    }

    public int ExpensiveProductsInStock()
    {
        List<Product> products = GetProductList();

        var expensiveInStockProducts = products.Where(p => p.UnitsInStock > 0
                                                           && p.UnitPrice > 3.00).ToList();

        Console.WriteLine("In-stock products that cost more than 3.00:");
        foreach (var product in expensiveInStockProducts)
            Console.WriteLine($"{product.ProductName} is in stock and costs more than 3.00.");
        return 0;
    }

    public int DisplayCustomerOrders()
    {
        List<Customer> customers = GetCustomerList();

        var waCustomers = customers.Where(c => c.Region == "WA").ToList();

        Console.WriteLine("Customers from Washington and their orders:");
        foreach (var customer in waCustomers)
        {
            Console.WriteLine($"Customer {customer.CustomerId}: {customer.CompanyName}");
            foreach (var order in customer.Orders)
                Console.WriteLine($"  Order {order.OrderId}: {order.OrderDate}");
        }

        return 0;
    }

    public int IndexedWhere()
    {
        string[] digits = { "zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine" };

        var shortDigits = digits.Where((digit, index) => digit.Length < index);

        Console.WriteLine("Short digits:");
        foreach (var d in shortDigits)
            Console.WriteLine($"The word {d} is shorter than its value.");

        return 0;
    }
}

File: Parallel-Java/src/parallel/linq101/Restrictions.java
package parallel.linq101;

import parallel.linq101.datasource.Customer;
import parallel.linq101.datasource.Data;
import parallel.linq101.datasource.Product;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class Restrictions {
    public List<Product> getProductList() {
        return Data.getProductList();
    }

    public List<Customer> getCustomerList() {
        return Data.getCustomerList();
    }

    public int lowNumbers() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        var lowNums = Arrays.stream(numbers).filter(n -> n < 5).toArray();
        System.out.println("Numbers < 5:");
        for (var x : lowNums)
            System.out.println(x);
        return 0;
    }

    public int productsOutOfStock() {
        List<Product> products = getProductList();
        var soldOutProducts = products.stream().filter(p -> p.getUnitsInStock() == 0)
                .collect(Collectors.toList());
        System.out.println("Sold out products:");
        for (var product : soldOutProducts)
            System.out.println(product.getProductName() + " is sold out!");
        return 0;
    }

    public int expensiveProductsInStock() {
        List<Product> products = getProductList();
        var expensiveInStockProducts = products.stream().filter(
                p -> p.getUnitsInStock() > 0 && p.getUnitPrice() > 3.00).collect(Collectors.toList());
        System.out.println("In-stock products that cost more than 3.00:");
        for (var product : expensiveInStockProducts)
            System.out.println(product.getProductName() + " is in stock and costs more than 3.00.");
        return 0;
    }

    public int displayCustomerOrders() {
        List<Customer> customers = getCustomerList();
        var waCustomers = customers.stream().filter(c -> c.getRegion() == "WA").collect(Collectors.toList());
        System.out.println("Customers from Washington and their orders:");
        for (var customer : waCustomers) {
            System.out.println("Customer " + customer.getCustomerId() + ": " + customer.getCompanyName());
            for (var order : customer.getOrders())
                System.out.println("  Order " + order.getOrderId() + ": " + order.getOrderDate());
        }
        return 0;
    }

    public int indexedWhere() {
        String[] digits = {"zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"};
        var shortDigits = IntStream.range(0, digits.length).filter(
                index -> digits[index].length() < index).mapToObj(index -> digits[index]).toArray();
        System.out.println("Short digits:");
        for (var d : shortDigits)
            System.out.println("The word " + d + " is shorter than its value.");
        return 0;
    }
}
